### PR TITLE
Updating parameter sweep docs

### PR DESCRIPTION
**Pull request by bknueven**
## Changes proposed in this PR
- [x] Break out optimize in RO_with_energy_recovery flowsheet & update parameter sweep how-to
- ~~Update RO_with_energy_recovery flowsheet & parameter sweep how-to to correctly set concentration~~ (Will wait for updates to NaCl Property Package)
- [x] Updates for parameter sweep docs to keep them up-to-date after #146.

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by bknueven** [comment]
I need to coordinate with @TimBartholomew on the second check box, but the new documentation for the changes to the parameter sweep tool could be reviewed.

**Comment by adam-a-a** [comment]
> I need to coordinate with @TimBartholomew on the second check box, but the new documentation for the changes to the parameter sweep tool could be reviewed.

@bknueven If I'm not mistaken, that second check box has to do with issue #57. Basically, we need to be able to define properties like volumetric flow rate and concentration and be able to have the state variables computed from those. Currently, we are restricted to always beginning with component mass flow rates (our state variables).

**Comment by bknueven** [comment]
Thanks for the clarification @adma-a-a. Given this is already captured, we can make a note on the issue.

File: docs/how_to_guides/how_to_use_parameter_sweep.rst [hunk 1]
```diff
@@ -69,11 +68,12 @@ Once this sequence of setup steps is performed, the parameters to be varied shou
...
+    sweep_params['Feed Mass NaCl'] = LinearSample(m.fs.feed.flow_mass_phase_comp[0, 'Liq', 'NaCl'], 0.005, 0.155, 4)
+    sweep_params['Water Recovery'] = LinearSample(m.fs.RO.recovery_mass_phase_comp[0, 'Liq', 'H2O'], 0.3, 0.7, 4)
 
-where the basic pattern is ``dict_name['Short/Pretty-print Name'] = (m.path.to.model.variable, lower_limit, upper_limit, num_samples)``.
-For example, "Feed Concentration", which is accessed in the model variable ``m.fs.feed.flow_mass_phase_comp[0, 'Liq', 'NaCl']``, is to be varied between 0.005 and 0.155 with 4 equally-spaced values, i.e., ``[0.005, 0.055, 0.105, 0.155]``.
+where the basic pattern is ``dict_name['Short/Pretty-print Name'] = LinearSample(m.path.to.model.variable, lower_limit, upper_limit, num_samples)``.
+For example, "Feed Mass NaCl", which is accessed in the model variable ``m.fs.feed.flow_mass_phase_comp[0, 'Liq', 'NaCl']``, is to be varied between 0.005 and 0.155 with 4 equally-spaced values, i.e., ``[0.005, 0.055, 0.105, 0.155]``.
+It is also possible to the parameter sweep tool to doing random sampling (uniform or normal), or for the user to specify their own sampling method.
```
Comment (adam-a-a) [inline docs/how_to_guides/how_to_use_parameter_sweep.rst:76]:
```suggestion
It is also possible to perform random sampling (uniform or normal) with the parameter sweep tool, or the user can specify their own sampling method.
```

File: docs/how_to_guides/how_to_use_parameter_sweep.rst [hunk 2]
```diff
@@ -69,11 +68,12 @@ Once this sequence of setup steps is performed, the parameters to be varied shou
...
-    sweep_params['Water Recovery'] = (m.fs.RO.recovery_mass_phase_comp[0, 'Liq', 'H2O'], 0.3, 0.7, 4)
+    sweep_params['Feed Mass NaCl'] = LinearSample(m.fs.feed.flow_mass_phase_comp[0, 'Liq', 'NaCl'], 0.005, 0.155, 4)
+    sweep_params['Water Recovery'] = LinearSample(m.fs.RO.recovery_mass_phase_comp[0, 'Liq', 'H2O'], 0.3, 0.7, 4)
 
-where the basic pattern is ``dict_name['Short/Pretty-print Name'] = (m.path.to.model.variable, lower_limit, upper_limit, num_samples)``.
-For example, "Feed Concentration", which is accessed in the model variable ``m.fs.feed.flow_mass_phase_comp[0, 'Liq', 'NaCl']``, is to be varied between 0.005 and 0.155 with 4 equally-spaced values, i.e., ``[0.005, 0.055, 0.105, 0.155]``.
+where the basic pattern is ``dict_name['Short/Pretty-print Name'] = LinearSample(m.path.to.model.variable, lower_limit, upper_limit, num_samples)``.
+For example, "Feed Mass NaCl", which is accessed in the model variable ``m.fs.feed.flow_mass_phase_comp[0, 'Liq', 'NaCl']``, is to be varied between 0.005 and 0.155 with 4 equally-spaced values, i.e., ``[0.005, 0.055, 0.105, 0.155]``.
```
Comment (adam-a-a) [inline docs/how_to_guides/how_to_use_parameter_sweep.rst:75]:
```suggestion
For example, "Feed Mass NaCl" (the feed mass flow rate of NaCl), which is accessed through the model variable ``m.fs.feed.flow_mass_phase_comp[0, 'Liq', 'NaCl']``, is to be varied between 0.005 and 0.155 with 4 equally-spaced values, i.e., ``[0.005, 0.055, 0.105, 0.155]``.
```

File: docs/technical_reference/tools/parameter_sweep.rst [hunk 3]
```diff
@@ -12,15 +12,28 @@ Overview
 The parameter sweep tool systematically fixes variables or modifies
 mutable parameters on a flowsheet (or Pyomo ConcreteModel), optimizes
 the flowsheet, and reports user-specified results.
-
-For each item the user wants to change, they specify a "short" name
-as a key, and then a tuple specifying the Pyomo object to be changed,
-a lower limit, an upper limit, and the number of elements to be sampled 
-for this parameter between the lower limit and upper limit.
+The parameter sweep tool can be operated in one of two ways: sweeping a 
+fixed set of paramters, or allowing for random samples from a distribution.
```
Comment (adam-a-a) [inline docs/technical_reference/tools/parameter_sweep.rst:16]:
```suggestion
fixed set of parameters, or allowing for random samples from a distribution.
```

File: docs/technical_reference/tools/parameter_sweep.rst [hunk 4]
```diff
@@ -12,15 +12,28 @@ Overview
...
+While different fixed or random sampling types can be combined, in a single
+parameter sweep the user must use either all fixed or all random sampling types.
```
Comment (adam-a-a) [inline docs/technical_reference/tools/parameter_sweep.rst:18]:
```suggestion
parameter sweep the user must use either all fixed or all random sampling types
```

File: docs/technical_reference/tools/parameter_sweep.rst [hunk 5]
```diff
@@ -12,15 +12,28 @@ Overview
...
+For the "fixed" sampling type, `LinearSample`, the parameter sweep tool will evaluate the 
+cross product of all the specified parameters, whereas with the "random" sampling
+types `UniformSample` and `NormalSample` the parameter sweep tool will evaluate
+a fixed number of samples specified in `num_samples`. With either sampling type,
+the values of the Pyomo objects (Vars, Expressions, etc.), that the user specifies
+in `outputs`.
+
+For each item the user wants to change, they specify a `sweep_params` dictionary.
+The keys are "short" names, and the values are one of the included `Sample` objects.
+In all cases the `Sample` objects are instancied with the Pyomo object to be changed,
```
Comment (adam-a-a) [inline docs/technical_reference/tools/parameter_sweep.rst:30]:
```suggestion
In all cases the `Sample` objects are instantiated with the Pyomo object to be changed,
```

File: docs/technical_reference/tools/parameter_sweep.rst [hunk 6]
```diff
@@ -12,15 +12,28 @@ Overview
...
-for this parameter between the lower limit and upper limit.
+The parameter sweep tool can be operated in one of two ways: sweeping a 
+fixed set of paramters, or allowing for random samples from a distribution.
+While different fixed or random sampling types can be combined, in a single
+parameter sweep the user must use either all fixed or all random sampling types.
+(This implementation detail may be relaxed in future releases.)
+
+For the "fixed" sampling type, `LinearSample`, the parameter sweep tool will evaluate the 
+cross product of all the specified parameters, whereas with the "random" sampling
+types `UniformSample` and `NormalSample` the parameter sweep tool will evaluate
```
Comment (adam-a-a) [inline docs/technical_reference/tools/parameter_sweep.rst:23]: @bknueven Is random sampling something we want to mention here, or should it be reserved for the subsequent how-to on parameter sweep that we discussed in #166? 

Either way, this shouldn't hold up the PR. If anything, a subsequent PR that adds a follow-up how-to for monte carlo sampling could describe random sampling types, and any undesired redundancies between the 2 how-to docs could then be eliminated.

**Review by adam-a-a — CHANGES_REQUESTED** [review]
Just some very minor changes suggested before merging

File: docs/technical_reference/tools/parameter_sweep.rst (hunk 4 above)
Comment (bknueven) [reply docs/technical_reference/tools/parameter_sweep.rst:18]: Why remove the period here? The parenthetical after this is its own sentence.

File: docs/technical_reference/tools/parameter_sweep.rst (hunk 6 above)
Comment (bknueven) [reply docs/technical_reference/tools/parameter_sweep.rst:23]: At the time I wrote this, #166 did not exist. I think it's fine for now to keep this an add on to #166 or further elaborate in subsequent documentation. At least with this the monte carlo stuff is at least mentioned.

File: docs/technical_reference/tools/parameter_sweep.rst (hunk 4 above)
Comment (adam-a-a) [reply docs/technical_reference/tools/parameter_sweep.rst:18]: No need to entertain this one. =)

**Review by adam-a-a — APPROVED** [review]

**Review by TimBartholomew — APPROVED** [review]
LGTM

**Pull request merged**